subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java

```diff
@@ -481,15 +482,31 @@ public void execute(DeleteSpec spec) {
     private void handleCollectedResults(TestCountLogger testCountLogger) {
         if (testCountLogger.hadFailures()) {
             handleTestFailures();
-        } else if (testCountLogger.getTotalTests() == 0 && shouldFailOnNoMatchingTests()) {
-            throw new TestExecutionException(createNoMatchingTestErrorMessage());
+        } else if (testCountLogger.getTotalTests() == 0) {
+            if (!hasFilter()) {
+                emitDeprecationMessage();
+            } else if (shouldFailOnNoMatchingTests()) {
```

Comment:
Good point. The default for `failOnNoMatchingTests` is currently `true` (see [here](https://github.com/gradle/gradle/blob/12f993ce40744abae39c13e656e482edcb10e82c/subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/filter/DefaultTestFilter.java#L34)) so we fail in the default case. This will not change when we deprecate "success on no test" in 9.0, and it is consistent with the no filter case where a deprecation warning is emitted. 

If `failOnNoMatchingTests == false` is configured, there is no deprecation warning and no exception thrown. If we find out there are no valid use cases for this, we can decide to deprecate this filter option, but I would suggest to wait and see before we decide this. FWIW, I found one usage of this configuration in our [docs project](https://github.com/gradle/gradle/blob/b09a4bb22666ca355be41b4314869a166002b0ac/subprojects/docs/build.gradle#L603).

Reply:
Ah, ok, for some reason I thought the default was false.  That all sounds sane and consistent then.  Just ignore me.